src/WindowsAzureMessaging/WindowsAzureMessaging/Helpers/MSInstallationManager.m

```diff
@@ -0,0 +1,174 @@
+
+@implementation MSInstallationManager
+
+- (instancetype)init {
+    if(self = [super init]) {
+        connectionDictionary = [MSInstallationManager parseConnectionString:_connectionString];
+        tokenProvider = [MSTokenProvider createFromConnectionDictionary:connectionDictionary];
+    }
+    
+    return self;
+}
+
++ (instancetype)sharedInstance {
+  dispatch_once(&onceToken, ^{
+    if (sharedInstance == nil) {
+      sharedInstance = [self new];
+    }
+  });
+  return sharedInstance;
+}
+
++ (void) initWithConnectionString:(NSString *)connectionString withHubName:(NSString *)hubName {
+    _connectionString = connectionString;
+    _hubName = hubName;
+}
+
++ (MSInstallation *) getInstallation {
+    return [[MSInstallationManager sharedInstance] getInstallation];
+}
+
++ (void) upsertInstallationWithDeviceToken: (NSString *) deviceToken {
+    [[MSInstallationManager sharedInstance] upsertInstallationWithDeviceToken:deviceToken];
+}
+
+- (MSInstallation *) getInstallation {
+    MSInstallation *installation = [MSLocalStorage loadInstallation];
+    
+    return installation;
+}
+
+- (void) upsertInstallationWithDeviceToken: (NSString *) deviceToken {
+    
+    if(!tokenProvider) {
+        NSLog(@"Invalid connection string");
+        return;
+    }
+        
+    MSInstallation *installation = [MSLocalStorage loadInstallation];
+    
+    if (!installation) {
+        installation = [MSInstallation createFromDeviceToken:deviceToken];
+    }
+    
+    NSString *endpoint = [connectionDictionary objectForKey:@"endpoint"];
+    NSString *url = [NSString stringWithFormat:@"%@%@/installations/%@?api-version=2017-04", endpoint, _hubName, installation.installationID];
+
+    NSString *sasToken = [tokenProvider generateSharedAccessTokenWithUrl:url];
+    NSURL *requestUrl = [NSURL URLWithString:url];
+    
+    NSDictionary *headers = @{
+       @"Content-Type" : @"application/json",
+       @"x-ms-version" : @"2015-01",
+       @"Authorization" : sasToken
+    };
+    
+    MSHttpClient *httpClient = [MSHttpClient new];
+
+    NSData *payload = [installation toJsonData];
+    
+    [httpClient sendAsync:requestUrl
+                method:@"PUT"
+                headers:headers
+                data:payload
+                completionHandler:^(NSData *responseBody, NSHTTPURLResponse *response, NSError *error) {
+        if (error) {
+            NSLog(@"Error via creating installation: %@", error.localizedDescription);
+        }
+        
+        [httpClient sendAsync:requestUrl method:@"GET" headers:headers data:nil completionHandler:^(NSData * _Nullable responseBody, NSHTTPURLResponse * _Nullable response, NSError * _Nullable error) {
+            NSString *str = [[NSString alloc] initWithData:responseBody encoding:NSUTF8StringEncoding];
+            [MSLocalStorage upsertInstallation:[MSInstallation createFromJsonString:str]];
+        }];
+    }];
+}
+
++ (NSDictionary*) parseConnectionString:(NSString*) connectionString
```

Comment:
Minor nit for this to have a space and then the brace on the same line, such as `if (foo) {` instead of new lines or without any space.  I'll note to get clang formatting set up for this